moto/glue/utils.py

```diff
@@ -0,0 +1,378 @@
+            return int(value)  # no size is enforced
+        except ValueError:
+            raise ValueError(f'"{value}" is not an integer.')
+
+    if type_ == "decimal":
+        try:
+            return float(value)
+        except ValueError:
+            raise ValueError(f"{value} is not a decimal.")
+
+    if type_ in ("char", "string", "varchar"):
+        return value  # no length is enforced
+
+    if type_ == "date":
+        try:
+            return datetime.strptime(value, "%Y-%m-%d").date()
+        except ValueError:
+            # NOTE AWS accepts up until 31st of every month, e.g. '2022-02-31'
+            warnings.warn("Date filtering beyond last of month not supported")
+            raise ValueError(f"{value} is not a date.")
+
+    if type_ == "timestamp":
+        match = re.search(
+            r"^(?P<timestamp>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})"
+            r"(?P<nanos>\.\d{1,9})?$",
+            value,
+        )
+        if match is None:
+            raise ValueError(
+                "Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]"
+                f" {value} is not a timestamp."
+            )
+
+        try:
+            timestamp = datetime.strptime(match.group("timestamp"), "%Y-%m-%d %H:%M:%S")
+        except ValueError:
+            # NOTE AWS accepts up until 31st of every month, e.g. '2022-02-31 00:00:00'
+            warnings.warn("Timestamp filtering beyond last of month not supported")
+            raise ValueError(
+                "Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]"
+                f" {value} is not a timestamp."
+            )
+
+        nanos = match.group("nanos")
+        if nanos is not None:
+            # strip leading dot, reverse and left pad with zeros to nanoseconds
+            nanos = "".join(reversed(nanos[1:])).zfill(9)
+            for i, nanoseconds in enumerate(nanos):
+                # NOTE precision loss here, as nanoseconds are not supported in datetime
+                microseconds = (int(nanoseconds) * 10**i) / 1000
+                if round(microseconds) == 0 and microseconds > 0:
+                    warnings.warn("Nanoseconds not supported, rounding to microseconds")
+
+                timestamp += timedelta(microseconds=round(microseconds))
+
+        return timestamp
+
+    raise InvalidInputException(
+        "An error occurred (InvalidInputException) when calling the"
+        f" GetPartitions operation: Unknown type : '{type_}'"
+    )
+
+
+class _Expr(abc.ABC):
+    @abc.abstractmethod
+    def eval(self, part_keys: List[Dict[str, str]], part_input: Dict[str, Any]) -> Any:
+        raise NotImplementedError()
+
+
+class _Ident(_Expr):
+    def __init__(self, tokens: ParseResults):
+        self.ident: str = tokens[0]
+
+    def eval(self, part_keys: List[Dict[str, str]], part_input: Dict[str, Any]) -> Any:
+        for key, value in zip(part_keys, part_input["Values"]):
+            if self.ident == key["Name"]:
+                try:
```

Comment:
There seems to be some duplication between `eval` and `leval` - is there a way to reduce this? Maybe by replacing this try-except with a call to `leval` instead?

Reply:
I thought about the same actually, but note how `eval` raises an `InvalidStateException`, while `leval` raises an `InvalidInputException`. This is because `eval` applies to existing partition values, and when an existing value cannot be cast to the schema (possibly because the schema changed after the initial value was written), an `InvalidStateException` is raised.

Correspondingly, when the user supplies a value that cannot be cast to the target partition column type, this is an `InvalidInputException`.

Maybe I can extract an `_eval` method and catch the `ValueError` separately.

Reply:
Yeah, I saw there were some differences, that's why I raised it as a question, instead of a request. If it doesn't work, or if the resulting code looks worse, than don't worry about it!